lib/util/result.ts

```diff
@@ -310,66 +310,68 @@ export class Result<T, E = Error> {
  *
  * All the methods resemble `Result` methods, but work asynchronously.
  */
-export class AsyncResult<T, E> extends Promise<Result<T, E>> {
-  constructor(
-    executor: (
-      resolve: (value: Result<T, E> | PromiseLike<Result<T, E>>) => void,
-      reject: (reason?: unknown) => void
-    ) => void
-  ) {
-    super(executor);
+export class AsyncResult<T, E> implements PromiseLike<Result<T, E>> {
+  private constructor(private asyncResult: Promise<Result<T, E>>) {}
+
+  then<TResult1 = Result<T, E>>(
+    onfulfilled?:
+      | ((value: Result<T, E>) => TResult1 | PromiseLike<TResult1>)
+      | undefined
+      | null
+  ): PromiseLike<TResult1> {
+    return this.asyncResult.then(onfulfilled);
   }
 
   static ok<T>(val: NonNullable<T>): AsyncResult<T, never> {
-    return new AsyncResult((resolve) => resolve(Result.ok(val)));
+    return new AsyncResult(new Promise((resolve) => resolve(Result.ok(val))));
   }
 
   static err<E>(err: NonNullable<E>): AsyncResult<never, E> {
-    return new AsyncResult((resolve) => resolve(Result.err(err)));
+    return new AsyncResult(new Promise((resolve) => resolve(Result.err(err))));
   }
 
   static wrap<T, E = Error, EE = never>(
-    promise: Promise<Result<T, EE>>
-  ): AsyncResult<T, E | EE>;
-  static wrap<T, E = Error>(
-    promise: Promise<NonNullable<T>>
-  ): AsyncResult<T, E>;
-  static wrap<T, E = Error, EE = never>(
-    promise: Promise<NonNullable<T> | Result<T, E | EE>>
+    promise: Promise<Result<T, EE>> | Promise<NonNullable<T>>,
+    onErr?: (err: NonNullable<E>) => Result<T, E>
   ): AsyncResult<T, E | EE> {
-    return new AsyncResult((resolve) => {
-      promise
-        .then((value) => {
+    return new AsyncResult(
+      promise.then<Result<T, E | EE>, Result<T, E | EE>>(
+        (value) => {
           if (value instanceof Result) {
-            return resolve(value);
+            return value;
           }
-
-          return resolve(Result.ok(value));
-        })
-        .catch((err) => resolve(Result.err(err)));
-    });
+          return Result.ok(value);
+        },
+        (err) => {
```

Comment:
this doesn't catch error from the then block, it only catches errors from existing promise